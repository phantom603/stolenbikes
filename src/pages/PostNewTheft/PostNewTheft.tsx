import React from 'react';
import * as Yup from 'yup';
import { Field, Formik } from 'formik';
import { TextField } from 'formik-material-ui';
import { Button, Grid } from '@material-ui/core';

import classes from './PostNewTheft.module.scss';

import UploadZone from '../../components/UploadZone';
import { postIncident } from '../../core/http/incident/incident.service';

import { NewIncidentType } from '../../core/interfaces/Incidents.type';

const errorMessage = {
  required: 'Field cannot be blank',
  stringTooLong: 'String is too long!',
  stringTooShort: 'String is too short!'
}

const PostNewTheft: React.FC = () => {
  const initialValues: NewIncidentType = {
    title: '',
    description: '',
    theftDate: '',
    reportedDate: '',
    location: '',
    files: []
  }

  const validationSchema = Yup.object().shape({
    title: Yup.string()
      .min(10, errorMessage.stringTooShort)
      .max(100, errorMessage.stringTooLong)
      .required(errorMessage.required),
    description: Yup.string()
      .min(10, errorMessage.stringTooShort)
      .max(250, errorMessage.stringTooLong)
      .required(errorMessage.required),
    theftDate: Yup.string()
      .required(errorMessage.required),
    reportedDate: Yup.string()
      .required(errorMessage.required),
    location: Yup.string()
      .max(100, errorMessage.stringTooLong)
      .required(errorMessage.required),
    files: Yup.array()
  })

  return (
    <div>
      <h1>Post new theft</h1>
      <Formik
        initialValues={initialValues}
        validationSchema={validationSchema}
        onSubmit={(values, {setSubmitting}) => {
          setSubmitting(false);
          postIncident(values)
        }}
      >
        {({handleSubmit, setFieldValue, isSubmitting}) => (
          <form onSubmit={handleSubmit}>
            <Grid container>
              <Grid item xs={12} md={6}>

                <Grid container spacing={2}>
                  <Grid item xs={12}>
                    <Field
                      component={TextField}
                      name="title"
                      label="Title"
                      maxLength="150"
                      variant="outlined"
                      className={classes.textField}
                      inputProps={{
                        maxLength: 100
                      }}
                    />
                  </Grid>
                  <Grid item xs={12}>
                    <Field
                      component={TextField}
                      name="description"
                      label="Description"
                      variant="outlined"
                      multiline
                      rows={5}
                      className={classes.textField}
                      inputProps={{
                        maxLength: 250
                      }}
                    />
                  </Grid>
                  <Grid item xs={12}>
                    <Field
                      component={TextField}
                      name="theftDate"
                      label="Date of the theft"
                      type="datetime-local"
                      variant="outlined"
                      className={classes.textField}
                      InputLabelProps={{
                        shrink: true,
                      }}
                    />
                  </Grid>
                  <Grid item xs={12}>
                    <Field
                      component={TextField}
                      name="reportedDate"
                      label="Reported date"
                      type="datetime-local"
                      variant="outlined"
                      className={classes.textField}
                      InputLabelProps={{
                        shrink: true,
                      }}
                    />
                  </Grid>
                  <Grid item xs={12}>
                    <Field
                      component={TextField}
                      name="location"
                      label="Location"
                      variant="outlined"
                      className={classes.textField}
                      inputProps={{
                        maxLength: 150
                      }}
                    />
                  </Grid>
                  <Grid item xs={12}>
                    <UploadZone onUploadCb={data => {
                      setFieldValue('files', data.map((file: File) => ({
                        fileName: file.name,
                        type: file.type,
                        size: file.size
                      })))
                    }}/>
                  </Grid>

                  <Grid item xs={12}>
                    <Button
                      variant="contained"
                      color="primary"
                      size="large"
                      disabled={isSubmitting}
                      type="submit"
                      className={classes.btn}
                    >
                      Post
                    </Button>
                  </Grid>
                </Grid>

              </Grid>
            </Grid>
          </form>
        )}
      </Formik>
    </div>
  )
}

export default PostNewTheft;