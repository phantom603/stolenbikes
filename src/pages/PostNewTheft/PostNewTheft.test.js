import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import PostNewTheft from './PostNewTheft';

const updateField = (wrapper, name, value) => {
  wrapper.simulate('change', {
    persist: () => {
    },
    target: {
      name,
      value,
    },
  })
}

describe('PostNewTheft component:', () => {
  test('renders without crashing', () => {
    shallow(<PostNewTheft/>);
  });

  test('should match stored snapshot', () => {
    const tree = renderer
      .create(<PostNewTheft/>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Title field:', () => {
  let tree;
  const inputSelector = 'input[name="title"]';

  beforeEach(() => {
    tree = mount(<PostNewTheft/>);
  })
  afterEach(() => {
    tree.unmount()
  })

  test('exists', () => {
    expect(tree.find(inputSelector)).toHaveLength(1)
  });

  test('should update an input when it is changed', () => {
    updateField(tree.find(inputSelector), 'title', 'Some bike title')

    const newValue = tree.find(inputSelector).props().value;

    expect(newValue).toEqual('Some bike title');
    expect(tree).toMatchSnapshot();
  });

  test('should show error with invalid title input field on blur', () => {
    const field = tree.find('input[name="title"]')
    field.simulate('focus')

    updateField(field, 'title', 'bad')
    field.simulate('blur', { target: { name: 'title', value: 'bad' } })

    tree.update()
    const updatedField = tree.find('input[name="title"]')
    expect(updatedField.props().value).toEqual('bad')
    expect(updatedField.parent().is('div')).toEqual(true)
  })

});
