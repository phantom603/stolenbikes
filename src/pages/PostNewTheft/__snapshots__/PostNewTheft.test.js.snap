// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostNewTheft component: should match stored snapshot 1`] = `
<div>
  <h1>
    Post new theft
  </h1>
  <form
    onSubmit={[Function]}
  >
    <div
      className="MuiGrid-root MuiGrid-container"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6"
      >
        <div
          className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <div
              className="MuiFormControl-root MuiTextField-root textField"
              maxLength="150"
            >
              <label
                className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                data-shrink={false}
              >
                Title
              </label>
              <div
                className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
                onClick={[Function]}
              >
                <input
                  aria-invalid={false}
                  autoFocus={false}
                  className="MuiInputBase-input MuiOutlinedInput-input"
                  disabled={false}
                  maxLength={100}
                  name="title"
                  onAnimationStart={[Function]}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  required={false}
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden={true}
                  className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    className="PrivateNotchedOutline-legendLabelled-3"
                  >
                    <span>
                      Title
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <div
              className="MuiFormControl-root MuiTextField-root textField"
            >
              <label
                className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                data-shrink={false}
              >
                Description
              </label>
              <div
                className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline"
                onClick={[Function]}
              >
                <textarea
                  aria-invalid={false}
                  autoFocus={false}
                  className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
                  disabled={false}
                  maxLength={250}
                  name="description"
                  onAnimationStart={[Function]}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  required={false}
                  rows={5}
                  value=""
                />
                <fieldset
                  aria-hidden={true}
                  className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    className="PrivateNotchedOutline-legendLabelled-3"
                  >
                    <span>
                      Description
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <div
              className="MuiFormControl-root MuiTextField-root textField"
            >
              <label
                className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined"
                data-shrink={true}
              >
                Date of the theft
              </label>
              <div
                className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
                onClick={[Function]}
              >
                <input
                  aria-invalid={false}
                  autoFocus={false}
                  className="MuiInputBase-input MuiOutlinedInput-input"
                  disabled={false}
                  name="theftDate"
                  onAnimationStart={[Function]}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  required={false}
                  type="datetime-local"
                  value=""
                />
                <fieldset
                  aria-hidden={true}
                  className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    className="PrivateNotchedOutline-legendLabelled-3 PrivateNotchedOutline-legendNotched-4"
                  >
                    <span>
                      Date of the theft
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <div
              className="MuiFormControl-root MuiTextField-root textField"
            >
              <label
                className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined"
                data-shrink={true}
              >
                Reported date
              </label>
              <div
                className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
                onClick={[Function]}
              >
                <input
                  aria-invalid={false}
                  autoFocus={false}
                  className="MuiInputBase-input MuiOutlinedInput-input"
                  disabled={false}
                  name="reportedDate"
                  onAnimationStart={[Function]}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  required={false}
                  type="datetime-local"
                  value=""
                />
                <fieldset
                  aria-hidden={true}
                  className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    className="PrivateNotchedOutline-legendLabelled-3 PrivateNotchedOutline-legendNotched-4"
                  >
                    <span>
                      Reported date
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <div
              className="MuiFormControl-root MuiTextField-root textField"
            >
              <label
                className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                data-shrink={false}
              >
                Location
              </label>
              <div
                className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
                onClick={[Function]}
              >
                <input
                  aria-invalid={false}
                  autoFocus={false}
                  className="MuiInputBase-input MuiOutlinedInput-input"
                  disabled={false}
                  maxLength={150}
                  name="location"
                  onAnimationStart={[Function]}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  required={false}
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden={true}
                  className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    className="PrivateNotchedOutline-legendLabelled-3"
                  >
                    <span>
                      Location
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <div>
              <div
                className="dropzone"
                onBlur={[Function]}
                onClick={[Function]}
                onDragEnter={[Function]}
                onDragLeave={[Function]}
                onDragOver={[Function]}
                onDrop={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                tabIndex={0}
              >
                <input
                  accept="image/*"
                  autoComplete="off"
                  multiple={false}
                  onChange={[Function]}
                  onClick={[Function]}
                  style={
                    Object {
                      "display": "none",
                    }
                  }
                  tabIndex={-1}
                  type="file"
                />
                <p>
                  Drag & drop some file here, or click to select file
                </p>
              </div>
              <ul />
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-contained btn MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge"
              disabled={false}
              onBlur={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="submit"
            >
              <span
                className="MuiButton-label"
              >
                Post
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`Title field: should update an input when it is changed 1`] = `ReactWrapper {}`;
